smolyay/grid.py

```diff
@@ -2,348 +2,429 @@
-        A set of unidimensional points and functions.
-
-    Raises
-    ------
-    TypeError
-        ``basis_set`` must be a :class:`BasisFunctionSet`.
+    domain: list
+         the domain of the grid
     """
 
-    def __init__(self, basis_set):
-        self.basis_set = basis_set
+    def __init__(self):
+        self._points = None
 
     @property
-    def basis_set(self):
-        """:class:`BasisFunctionSet`: basis set (points and functions)."""
-        return self._basis_set
-
-    @basis_set.setter
-    def basis_set(self, basis_set):
-        if not isinstance(basis_set, BasisFunctionSet):
-            raise TypeError('Basis set must be a BasisFunctionSet')
-        self._basis_set = basis_set
+    @abc.abstractmethod
+    def dimension(self):
+        """int: number of independent variables."""
+        pass
 
+    @property
     @abc.abstractmethod
-    def __call__(self, dimension):
-        """Make grid points and their corresponding basis functions.
+    def domain(self):
+        """numpy.ndarray: domain of the point set."""
+        pass
+
+    @property
+    def points(self):
+        """numpy.ndarray: Points to be sampled."""
+        if self._points is None:
+            self._generate_points()
+        return self._points
+
+    @staticmethod
+    def _scale_to_domain(x, old, new):
+        """Transform the points into new domain.
 
         Parameters
         ----------
-        dimension: int
-            Number of idependent variables.
+        x: list
+            point(s) to be transformed.
+        old: list
+            old domain.
+        new: list
+            new domain.
 
         Returns
         -------
-        :class:`IndexGrid`
-            Grid points (and their indexes) and their corresponding functions.
+        numpy.ndarray
+            Transformed point(s).
+
+        Raises
+        ------
+        TypeError
+            Old and new domain must have the same shape.
+            Domain should be a dim x 2 array.
         """
-        pass
+        old = numpy.array(old, copy=False, ndmin=2)
+        new = numpy.array(new, copy=False, ndmin=2)
+
+        # error checking
+        if old.shape != new.shape:
+            raise TypeError("Old and new domain must have the same shape")
+        if old.ndim != 2 or old.shape[1] != 2:
+            raise TypeError("Domain should be a dim x 2 array")
+
+        new_x = new[:, 0] + (new[:, 1] - new[:, 0]) * (
+            (x - old[:, 0]) / (old[:, 1] - old[:, 0])
+        )
+        # clamp bounds
+        if len(new_x.shape) == 1:
+            numpy.clip(new_x, new[:, 0], new[:, 1], out=new_x)
+        else:
+            for i in range(new.shape[0]):
+                numpy.clip(new_x[..., i], new[i, 0], new[i, 1], out=new_x[..., i])
+        return new_x
+
+    def set_params(self, **parameters):
+        for parameter, value in parameters.items():
+            setattr(self, parameter, value)
+        self._generate_points()
+        return self
```

Comment:
It is better to have actual attributes (as properties), rather than use a catch all method like this.
```suggestion
```